### PR TITLE
Cosine annealing with restarts

### PR DESCRIPTION
**Pull request by striajan**
I implemented cosine annealing with warm restarts. The implementation is stateless, computing number of already performed restarts and epoch of the last restart directly from the `self.last_epoch` variable. This resolves issues in the implementations #6130 and #7821. It was inspired by TensorFlow [implementation](https://github.com/tensorflow/tensorflow/blob/25c197e02393bd44f50079945409009dd4d434f8/tensorflow/python/training/learning_rate_decay.py#L514), but it was implemented in more readable style.

There is one important difference compared to the original [paper](https://arxiv.org/abs/1608.03983). The authors let the variable T_cur go from 0 to T_i (i-th run has T_i + 1 iterations). In my implementation, T_cur goes from 0 to T_i - 1 (i-th run has T_i iterations only). This is more consistent with `StepLR`, where the decay happens at `step_size`. It is also more convenient to use T_max (and T_i computed from it) to define length of the period between restarts. This approach is used in TensorFlow as well. The difference is that they use T_cur / T_i instead of T_cur / (T_i - 1), therefore never reaching the minimum LR. If the TensorFlow approach is more preferred, I can update my code easily.

I have also included a possibility of decaying the initial LR after each restart by the multiplicative factor gamma. This was inspired by TensorFlow implementation. The paper doesn't mention this explicitly, although it defines the minimum and maximum eta for the i-th run.

Here are some example plots of generated LR schedules. It might be nice to have them in the documentation, as we do have plots for various activation functions. I can add the generating script to docs/src/scripts if desired.

cosineannealingrestartslr

I implemented the schedule in a separate class `CosineAnnealingRestartsLR` to avoid possible inconsistencies with `CosineAnnealingLR`, which were mentioned in #6130. TensorFlow has two independent implementations as well.

I am not completely happy with naming of variables. In my opinion, `eta_min` should be named `lr_min`, since PyTorch uses `lr` and `base_lr`. I also don't like names `T_max`, `T_mult` and `T_i` which violate [PEP 8](https://www.python.org/dev/peps/pep-0008/). I used these names to be consistent with `CosineAnnealingLR` and with the paper. I can update them if desired.

### CODE REVIEW

**Comment by striajan** [comment]
I have added missing tests. Namely test targeting the second base learning rate 0.5 and test of state dictionary. I have also simplified the testing code a bit.

**Comment by striajan** [comment]
I also want to comment on the different implementations of the [equation 5](https://arxiv.org/pdf/1608.03983.pdf). Let me remind that in the paper T_cur iterates from 0 to T_i in each run. In my and TensorFlow implementation it iterates from 0 to T_i - 1. In TensorFlow they divide T_cur / T_i. In my code I use T_cur / (T_i - 1) to reach eta_min in the last iteration of each run. However, since eta_min defaults to 0, the gradients don't get updated in the last iteration. Therefore it might be better to use TensorFlow approach.

**Comment by striajan** [comment]
In the current version of doc-string, I use i to denote run between (i - 1)-th and i-th restart, i.e. i starts from 1, which puts i - 1 to the exponents. Maybe it would be better to start with run 0. Moreover, the epochs in PyTorch start with 0 as well. The parameter `T_max` could be then called `T_0`. Please let me know.

**Comment by striajan** [comment]
I was thinking about my last two comments and have found them reasonable. I did the following chages:
- I use T_cur / T_i in [equation 5](https://arxiv.org/pdf/1608.03983.pdf), never reaching (possibly zero) eta_min. Now the behavior is consistent with TensorFlow.
- I renamed `T_max` to `T`. The name is therefore inconsistent with `CosineAnnealingLR`, however, it also has different meaning. There is no actual maximum in `CosineAnnealingRestartsLR`. This is also a valid point for keeping the classes separated.
- The variable i in the docstring now denotes both the number of performed restarts and the current run, i.e. the initial run is i = 0. This is consistent with numbering of epochs.

I updated the example plots:
cosineannealingrestartslr

The PR is now ready for the review.

**Comment by loshchil** [comment]
@striajan Thanks a lot for your contribution! BTW, the TensorFlow code that you mentioned was implemented by @PatrykChrabaszcz and the first author of the paper, so it is good that you used that implementation as a reference.

**Comment by ahundt** [comment]
this has been idle for a while could you consider updating / merging it?

**Comment by striajan** [comment]
I merged my PR with recent master. I noticed that the logic behind updating the learning rate has changed meanwhile - it is evaluated recursively #14010. This is probably not possible for `CosineAnnealingRestartsLR`, as the restarted LR is derived from the base LR. Therefore it works like `LambdaLR` and is not consistent with `StepLR`, `MultiStepLR`, `ExponentialLR`, `CosineAnnealingLR`.

**Comment by striajan** [comment]
I found a way how to compute restarted LR recursively. If it needs to be restarted, the initial LR for the previous run is computed using LR from the previous epoch (i.e. the last epoch of the previous run). Then it can be used to compute the initial LR for the current run. If no restart is needed, the LR is updated similarly as in `CosineAnnealingLR`.

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -265,6 +267,88 @@ def get_lr(self):
...
+        else:
+            # computation of the last restarting epoch is based on sum of geometric series:
+            # last_restart = T * (1 + T_mult + T_mult ** 2 + ... + T_mult ** i_restarts)
+            i_restarts = int(math.log(1 - self.last_epoch * (1 - self.T_mult) / self.T,
+                                      self.T_mult))
+            last_restart = int(self.T * (1 - self.T_mult ** i_restarts) / (1 - self.T_mult))
+
+        if self.last_epoch == last_restart:
+            T_i1 = self.T * self.T_mult ** (i_restarts - 1)  # T_{i-1}
+            lr_update = self.eta_mult / self._decay(T_i1 - 1, T_i1)
```
Comment (KarimTarabishy) [inline torch/optim/lr_scheduler.py]: shouldn't it be a multiplication? This would produce values larger than 1.
